mitiq/pt/pt.py

```diff
@@ -0,0 +1,61 @@
+# Copyright (C) Unitary Fund
+#
+# This source code is licensed under the GPL license (v3) found in the
+# LICENSE file in the root directory of this source tree.
+
+from typing import Callable, List, Optional, Tuple, Union
+
+import cirq
+import numpy as np
+
+from mitiq import QPROGRAM, Executor, Observable, QuantumResult
+from mitiq.interface import atomic_converter
+
+
+def execute_with_pauli_twirling(
+    circuit: QPROGRAM,
+    executor: Union[Executor, Callable[[QPROGRAM], QuantumResult]],
+    observable: Optional[Observable] = None,
+    *,
+    num_circuits: int = 100,
+) -> float:
+    executor = (
+        executor if isinstance(executor, Executor) else Executor(executor)
+    )
+    twirled_circuits = pauli_twirled_circuits(circuit, num_circuits)
+    expvals = executor.evaluate(twirled_circuits, observable)
+    return np.average(expvals)
+
+
+def pauli_twirled_circuits(
+    circuit: QPROGRAM, num_circuits: int
+) -> List[QPROGRAM]:
+    return [add_paulis(circuit) for _ in range(num_circuits)]
+
+
+Paulis = Union[cirq.IdentityGate, cirq.XPowGate, cirq.YPowGate, cirq.ZPowGate]
+
+
+def sample_paulis() -> Tuple[Paulis, Paulis, Paulis, Paulis]:
+    # P, Q, R, S
+    return cirq.ops.X, cirq.ops.I, cirq.ops.X, cirq.ops.X
+
+
+@atomic_converter
+def add_paulis(circuit: cirq.Circuit) -> cirq.Circuit:
+    return circuit.map_operations(twirl_CNOT)
+
+
+def twirl_CNOT(op: cirq.Operation) -> cirq.OP_TREE:
+    if op.gate != cirq.CNOT:
+        return op
+
+    P, Q, R, S = sample_paulis()
```

Comment:
I guess `sample_paulis()` will sample a single pair Paulis and deduce the other pair assuming they will be applied to a CNOT gate.

It would be conceptually a bit clearer if such "deduction" happened inside `twirl_CNOT` such that `sample_paulis()` can simply return 2 to random pauli gates.

Reply:
Ah I was thinking of a slightly more brute force approach, which is how I interpret what happens in this [reference](https://arxiv.org/pdf/2301.02690.pdf) which is to choose $P, Q, R, S$ randomly from the following table.

<img width="441" alt="Screenshot 2023-04-26 at 6 30 35 AM" src="https://user-images.githubusercontent.com/12703123/234591146-3b8b085b-af51-414d-991a-44b91c932439.png">

This is exactly what you suggest, but pre-computed. Do you have a preference? I'm not actually sure how I would go about computing $R$ and $S$ given $P$ and $Q$ if we take that route. Do you have ideas there?